### PR TITLE
Fix validation when there are undefined entities (#539)

### PR DESCRIPTION
**Pull request by ghost**
(c) Toms/Frank :)

### CODE REVIEW

**Comment by ghost** [comment]
The CI error here is this one:
>ERROR:Attempt to load network entity http://www.oasis-open.org/docbook/xml/4.5/dbcentx.mod (line 0 column 0)

It's a bit confusing. ~~I looked at [Ubuntu packages](https://packages.ubuntu.com/search?keywords=xml+catalog&searchon=all&suite=hirsute&section=all) and it appears that XML catalogs need the `xml-core` package.~~ That package is definitely installed in the CI container. The dbcentx.mod file can even literally be downloaded from the address in the error message. And GHA definitely has web access, so even if the catalog is effed, we should not be seeing this message, imo. Ideas, @tomschr ?

I have checked more better and noticed that the actual `xmlcatalog` tool is contained in `libxml2-utils` which is currently ~~not~~ _also_ installed.

**edits** xml-core is a red herring, libxml2-utils the actual package and it's also installed.

**Comment by ghost** [comment]
<details>
<summary>Running tests locally on Leap 15.3, I get this</summary>

```
daps/test > ./run_tests.sh --images --xsltprocessor=xsltproc

############################################################
#       Testing DAPS 2021-07-29T17:40:28 CEST         #
############################################################
#
# XSLT Processors: xsltproc
#

===========================================================
= xsltproc 
===========================================================

Test: Document Source Validation
==================================================
test_validSources

Ran 1 test.

OK
....................

Test: Image Generation
==================================================
test_images
test_imagesDIA
test_imagesJPG
test_imagesODG
test_imagesDITAA
test_imagesPNG
test_imagesSVG
test_imagesIGNORE
test_imagesROOTID

Ran 9 tests.

OK
....................

--------------------------------------------------
Overall Statistics

Tests Total:   10
Tests Passed:  10
Tests Skipped: 0
Tests Failed:  0
```

</details>

This looks ok to me. However, it did go much _much_ faster than the 15 minutes that @fsundermeyer said it would (it literally took a few seconds only).

File: libexec/daps-xmlwellformed [hunk 1]
```diff
@@ -1,172 +1,156 @@
...
 #
-# This script is needed as xmllint cannot ignore double IDs
-# lxml
-#
-#
 """Performs a well-formedness check on XML.
 
- * Does XInclude processing before any checks;
- * Warns about files which cannot be found
+This script is needed as xmllint cannot ignore double IDs.
```
Comment (ghost) [inline libexec/daps-xmlwellformed]: Nit, "double IDs" is bad English:

```suggestion
This script is needed as xmllint cannot ignore duplicated IDs.
```

File: bin/daps.in [hunk 2]
```diff
@@ -1725,6 +1726,21 @@ if [[ "adoc" = "$SRC_FORMAT" ]]; then
 
     make $MOPTS -f $DAPSROOT/make/adoc2xml.mk  || exit_on_error "\nConverting the AsciiDoc sources to XML caused a fatal error.\nCheck the $(basename ${ASCIIDOC}) message above for details."
 fi
+
+#-------------------------------------------------------------------
+# Check whether the documents are _well-formed_ (not if valid) - if not,
+# exit and display the error message
+# Works for both, DocBook4 and DocBook5 since we are only checking for
+# well-formdness and not for validity (a DocBook5 validity check would
+# require jing)
```
Comment (ghost) [inline bin/daps.in]: Wording nits:

```suggestion
# Check whether the documents are _well-formed_ (but not whether they
# are valid). If it is not, exit and display the error message.
# Works for both DocBook4 and DocBook5, as we are only checking for
# well-formedness and not for validity (a DocBook5 validity check would
# require jing).
```

File: libexec/daps-xmlwellformed [hunk 3]
```diff
@@ -1,172 +1,156 @@
...
+It performs:
+
+ * Does XInclude processing if requested (option --xinclude)
  * Ignores non-unique IDs (attributes xml:id or id)
+ * Output errors about undeclared entities
```
Comment (ghost) [inline libexec/daps-xmlwellformed]:
```suggestion
This tool:

 * Processes XIncludes (if requested using the option --xinclude)
 * Ignores non-unique IDs (attributes xml:id or id)
 * Outputs errors about undeclared entities
```

File: libexec/daps-xmlwellformed [hunk 4]
```diff
@@ -1,172 +1,156 @@
...
-
-if etree.LXML_VERSION < (3, 4, 0):
-    print("ERROR: I need a minimum version of 3.4.0 of lxml.",
-          file=sys.stderr)
+ROOTLOGGER = "lxmlerrors"
+# Hint: file keyword needs to be filled with extra keyword on the
+# logger methods .critical, .debug(), .error(), .fatal(), and .warn().
+# For example:
+#   log.error("...", extra=dict(file="foo.xml", ))
+# Basically everything can be used what is in ERROR_PATTERN
```
Comment (ghost) [inline libexec/daps-xmlwellformed]: ~~This sentence is definitely bad English but I don't quite understand what it means, so don't know how I would reword.~~ Let's ignore this for the moment.

File: libexec/daps-xmlwellformed [hunk 5]
```diff
@@ -208,27 +192,42 @@ def parse_cli(args=None) -> argparse.Namespace:
...
-if __name__ == "__main__":
-    args = parse_cli()
-    endresult=0
-    failures=0
-    successes=0
+def main(cliargs=None) -> int:
+    """Entry point for the application script
+
+    :param list cliargs: Arguments to parse or None (=use :class:`sys.argv`)
+    :return: error code; 0 => everything was succesfull, !=0 => error
```
Comment (ghost) [inline libexec/daps-xmlwellformed]:
```suggestion
    :return: error code; 0 => everything was successful, !=0 => error
```

**Review by ghost — COMMENTED** [review]
With the reproducers I posted in #539, this now looks good to me! Thanks a lot. I am all in favor of merging this one.

```
> ddaps -d DC-SLES-all validate
Fatal error:
/home/sknorr-l/data/gits/doc-sle/xml/smt_product_overview.xml:
    ERROR:Entity 'ismt' not defined (line 24 column 10)
```

I am not going to pretend I properly reviewed the entire Python code though. Seems as though it works, however. Below are just typos in comments and the like. Nothing functional.

File: make/common_variables.mk [hunk 6]
```diff
@@ -257,16 +251,34 @@ DAPSSTRINGS := --stringparam "converter.name=$(MY_NAME)" \
...
 #
-IMG_DIRECTORIES := $(IMG_GENDIR)/gen/png $(IMG_GENDIR)/gen/pdf \
+IMG_GENDIR         := $(BUILD_DIR)/.images
+ifneq "$(strip $(ADOC_IMG_DIR))" ""
+  IMG_SRCDIR := $(BUILD_DIR)/.adoc_images/src
+else
+  IMG_SRCDIR         := $(DOC_DIR)/images/src
+endif
+
+IMG_GEN_DIRECTORIES := $(IMG_GENDIR)/gen/png $(IMG_GENDIR)/gen/pdf \
```
Comment (ghost) [inline make/common_variables.mk]: I think you removed PDF image generation, no?

```suggestion
IMG_GEN_DIRECTORIES := $(IMG_GENDIR)/gen/png \
```

File: make/validate.mk [hunk 7]
```diff
@@ -89,6 +93,13 @@ endif
 	@exit 1
     endif
   endif
+  ifeq "$(strip $(VALIDATE_TABLES))" "1"
+    ifneq "$(strip $(FAULTY_TABLES))" ""
+	@ccecho "warn" "Fatal error: The following tables contain errors:"
```
Comment (ghost) [inline make/validate.mk]: I am wondering if I am misunderstanding the meaning of "fatal": This is a "warn"-type ccecho & the --help says "issues a non-fatal warning" but the message text reads "Fatal error: ..." & the exit code I see is `2`.

What I'd expect is getting a message + a non-zero error code:
* Being fatal, it is more useful for CI.
* Some errors detected by the table tool break FOP PDF builds.

It appears to deliver on that promise already, but the wording should be made consistent.

**Review by ghost — COMMENTED** [review]
1. I merged three (previously four) commits to `main` already, namely the entity-issue fixes and the `fop -version` fix. Commits that I either had a question about or where it seemed they depended on a previous commit are still on this branch.
2. **I force-pushed this branch**, so it's now up-to-date with main. Before you push to this branch again, make sure you have no uncommitted/unpushed work on the branch and then `git fetch; git reset --hard origin/fix_validation` (Slightly evil, I know.)
3. Additional comments after the break

---

* All new features/fixes work for me. Thanks!

* I would probably support making table validation a mandatory part of validation, rather than hiding it behind a parameter. I ran with and without `--validate-tables` under `time`. For DC-SLES-all, I see 21s (with --validate-tables) and 19s (without), and it varies a bit. That difference almost negligible and given the value that table validation brings, it's imo best to enable it by default. I also checked what the situation looked like before Carla replaced the tables in the AY guide, and times are *very* similar.
Imo, the only reason to not make integrate into normal validation was if we expected the script to have bugs/false positives. However, we've been running it in CI for over a year and I have not seen issues, except that it's maybe a bit strict.

* From what I can see, if DAPS runs into an issue in --validate-ids, --validate-images, or --validate-tables, it will just stop without executing the other options. However, if a document independently contains duplicated images, a bad table and underscores in IDs, the user will have to validate three times before everything is fixed. If they trust in CI a little too much, they may even push three times and waste CI cycles that way. Either way, it's a slightly annoying experience.
Can you make these options execute independently of whether any of the others has failed already?
(If that's too big for this PR, I can open an issue and it can be handled later.)

File: make/validate.mk (hunk 7 above)
Comment (fsundermeyer) [reply make/validate.mk]: Addressed in 0d93029

**Comment by ghost** [comment]
@fsundermeyer Thanks a lot! I just did a quick check for the new validation fixes before merging but this really is awesome!
